### PR TITLE
[WIP] selective functor

### PR DESCRIPTION
**Pull request by richard-gibson**
for #1329
continuing with unfinished tasks, feedback appreciated
* [X]  Interface definition
  * [X] A couple of operators, like the OCaml version maybe
   * [X] select
   * [X] branch
   * [X] whenS
   * [X] ifS
   * [X] orS
   * [X] andS
   * [ ] whileS
   * [ ] fromMaybe
   * [ ] anyS
   * [ ] allS
  
* [X]  Update Monad
* [X]  Laws
* [ ]  Instances
  
  * [X]  Option (prob can be removed as we now use Option.monad())
  * [X]  Identity
  * [X]  Func0
  * [ ]  https://github.com/snowleopard/selective/blob/master/src/Control/Selective.hs#L311-L321
* [X]  Tests
* [ ]  Documentation

### CODE REVIEW

File: modules/core/arrow-core-data/src/test/kotlin/arrow/core/OptionTest.kt [hunk 1]
```diff
@@ -49,7 +50,8 @@ class OptionTest : UnitSpec() {
       TraverseFilterLaws.laws(Option.traverseFilter(), Option.applicative(), ::Some, Eq.any()),
       MonadFilterLaws.laws(Option.monadFilter(), ::Some, Eq.any()),
       HashLaws.laws(Option.hash(Int.hash()), Option.eq(Int.eq())) { it.some() },
-      SemigroupalLaws.laws(Option.semigroupal(), ::Some, ::bijection, associativeSemigroupalEq)
+      SemigroupalLaws.laws(Option.semigroupal(), ::Some, ::bijection, associativeSemigroupalEq),
+      SelectiveLaws.laws(Option.selective(), Eq.any())
```
Comment (pakoito) [inline modules/core/arrow-core-data/src/test/kotlin/arrow/core/OptionTest.kt]: This is not necessary as MonadFilterLaws already check SelectiveLaws. Same for the others.

File: modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/function0.kt [hunk 2]
```diff
@@ -42,6 +43,12 @@ interface Function0Applicative : Applicative<ForFunction0> {
     Function0.just(a)
 }
 
+@extension
+interface Function0Selective : Selective<ForFunction0>, Function0Applicative {
+  override fun <A, B> Function0Of<Either<A, B>>.select(f: Kind<ForFunction0, (A) -> B>): Kind<ForFunction0, B> =
+    fix().flatMap  { it.fold({l -> just(l).ap(f)}, {r -> just(identity(r))})}
```
Comment (pakoito) [inline modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/function0.kt]: Typo: additional whitespace. Format it with IntelliJ.

File: modules/core/arrow-test/src/main/kotlin/arrow/test/laws/SelectiveLaws.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import arrow.typeclasses.Eq
+import arrow.typeclasses.Selective
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+
+object SelectiveLaws {
+
+  fun <F> laws(A: Selective<F>, EQ: Eq<Kind<F, Int>>): List<Law> =
+    ApplicativeLaws.laws(A, EQ) + listOf(
+      Law("Selective Laws: identity") { A.identityLaw(EQ) },
```
Comment (pakoito) [inline modules/core/arrow-test/src/main/kotlin/arrow/test/laws/SelectiveLaws.kt]: Did they have more descriptive names on the paper?

File: modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/option.kt [hunk 4]
```diff
@@ -123,6 +123,12 @@ interface OptionApplicative : Applicative<ForOption> {
     Option.just(a)
 }
 
+@extension
+interface OptionSelective : Selective<ForOption>, OptionApplicative {
+  override fun <A, B> OptionOf<Either<A, B>>.select(f: OptionOf<(A) -> B>): Option<B> =
+    fix().flatMap  { it.fold({l -> just(l).ap(f)}, {r -> just(identity(r))})}
```
Comment (pakoito) [inline modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/option.kt]: Same whitespace

**Comment by pakoito** [comment]
>UnusedImports - [IdTest.kt] at /home/travis/build/arrow-kt/arrow/modules/core/arrow-core-data/src/test/kotlin/arrow/core/IdTest.kt:20:1

Run `detekt` locally I say.

File: modules/core/arrow-core-data/src/test/kotlin/arrow/core/OptionTest.kt (hunk 1 above)
Comment (richard-gibson) [reply modules/core/arrow-core-data/src/test/kotlin/arrow/core/OptionTest.kt]: Cool, will remove. In that case I should remove all selective instances if there is a monad instance available too

File: modules/core/arrow-test/src/main/kotlin/arrow/test/laws/SelectiveLaws.kt (hunk 3 above)
Comment (richard-gibson) [reply modules/core/arrow-test/src/main/kotlin/arrow/test/laws/SelectiveLaws.kt]: Distributivity, Associativity, selectM == select if a Monad. 
I can add.

**Pull request merged**